Add support for has("wsl") to detect Windows Subsystem for Linux

mqudsi opened the pull request:
Based on the value of `CMAKE_SYSTEM_VERSION` (which is the output of
`uname -r` where available), detect when running under WSL and define
a preprocessor variable WSL. Use that to set the feature flag `has("wsl")`.

Closes #7329

mqudsi committed:
Add support for has("wsl") to detect Windows Subsystem for Linux

Based on the value of CMAKE_SYSTEM_VERSION (which is the output of
`uname -r` where available, detect when running under WSL and define
a preprocessor variable WSL.

Closes #7329

jamessan commented on src/nvim/eval.c:
```diff
@@ -10603,6 +10603,9 @@ static void f_has(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     "winaltkeys",
     "writebackup",
+#if defined(WSL)
+    "wsl",
```
This would also need to be documented.

jamessan commented on CMakeLists.txt:
```diff
@@ -44,6 +44,11 @@ if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
   endif()
 endif()
 
+if (CMAKE_HOST_SYSTEM_VERSION MATCHES ".*-Microsoft")
```
This won't match for non-WSL builds on Windows?

mqudsi replied:
According to the CMAKE docs, it's only available on platforms with uname: https://cmake.org/cmake/help/v3.4/variable/CMAKE_HOST_SYSTEM_VERSION.html?highlight=uname

> A numeric version string for the system. On systems that support uname, this variable is set to the output of uname -r. On other systems this is set to major-minor version numbers.

On regular Windows that would 6.1

justinmk commented:
Let's document our assumption in `test/functional/eval/has_spec.lua` :

```
  it('"wsl"', function()
    if 1 == funcs.has('win32') then
      eq(0, funcs.has('wsl'))
    end
  end)
```

Shougo commented:
Why the documentation is not updated?

mqudsi commented:
Just an update, I'm out of the country for a while and likely won't be able to get back to this until the second week of October.

@justinmk thanks, that looks good. I'll add that.

@Shougo no reason whatsoever besides not knowing which documentation should be updated. Any particular places you'd like to see this reflected?

Shougo commented:
> @Shougo no reason whatsoever besides not knowing which documentation should be updated. Any particular places you'd like to see this reflected?

`:help feature-list` should be updated.

justinmk commented:
@mqudsi ping :)

mqudsi committed:
Add sanity checks to make sure WSL isn't matched incorrectly

Assert that WSL is never defined simultaneously with mac or win32, lua
code per #7330 by @justinmk.

mqudsi committed:
Include `has("wsl")` in `:help feature-list`

mqudsi commented:
I completely dropped the ball on this, but I've just rebased against master and added the assertions and suggested documentation. Any other things I need to do before this can be merged?

justinmk commented on CMakeLists.txt:
```diff
@@ -44,6 +44,11 @@ if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
 
+if (CMAKE_HOST_SYSTEM_VERSION MATCHES ".*-Microsoft")
+  # Detect WSL, which is also UNIX but not Windows/WIN32/WIN64
+  add_definitions(-DWSL)
```
"WSL" for a build-time define seems like it could risk conflicting with some other C library define. Let's qualify it a bit, e.g. `WIN32_WSL`.

justinmk commented:
@mqudsi  I assume you tested this already on an actual WSL system?

Will merge after above comments are addressed.

mqudsi committed:
Rename global define WSL to NVIM_WSL for #7330

Prevent possible collision with 3rd party modules/sources also using WSL
to denote the presence of the Windows Subsystem for Linux.

mqudsi commented:
@justinmk Yup. Tested under both FreeBSD (to make sure it returns 0) and WSL (where it returns 1). 

The rename from `WSL` to `NVIM_WSL` has been committed to the branch, btw.

mqudsi committed:
fixup! Rename global define WSL to NVIM_WSL for #7330

justinmk commented:
Merged, thanks.

I changed the symbol to `HAVE_WSL` and moved it to `config/CMakeLists.txt` since that is the usual pattern for this type of thing.

mqudsi commented:
Awesome, thanks!

grolongo commented:
Fresh install of Neovim 0.3.0 from Debian Unstable repository in Debian WSL.

`:echo has("wsl")` always returns 0

am I missing something?

justinmk commented:
Did you build from source? What does CMAKE_HOST_SYSTEM_VERSION say if you print it from cmake ?

grolongo commented:
I didn't build from source and don't have cmake installed by default on WSL so I can't tell.  
I installed using the unstable Debian repo (where neovim is updated to 0.3.0).  
I'm pretty sure the maintainer didn't use WSL to build it and therefore it doesn't work.

mqudsi commented:
@justinmk @sunilmt @therealkenc do you have any thoughts on how to best handle this? At build-time it is possible to use `uname` at the command line to set a `constexpr` value or `#define` that can enable detection of Windows, but the binaries installed via `apt` are the same ones available on real Ubuntu and not compiled under WSL, so detection fails.

Is a call to `uname(2)` (so in code, not in a shell) the only (and most efficient) way to go here?

Facing the same issue [in fish](https://github.com/fish-shell/fish-shell/issues/5115#issuecomment-408640314), too.
